### DISCUSSION TITLE
yarn generate gets stuck on something

### DISCUSSION THREAD

**Discussion by ZaneH** [prisma/prisma#4038] [Q&A] [answered]
Hi, I'm having what feels like a simple issue, but I can't find a fix.

First, I have this TypeScript module that simply loads an Ethereum wallet from an environment variable.
```ts
// ./modules/Wallet.ts
import Web3 from 'web3'

// create a Web3 connection
export const web3 = new Web3(
  process.env.NODE_ENV === 'production'
    ? 'wss://mainnet.infura.io/ws/v3/redacted'
    : 'wss://ropsten.infura.io/ws/v3/redacted'
)

// load an Ethereum wallet from a private key (env var)
const envWallet = () => {
  if (!process.env.ETH_PRIVATE_KEY) {
    require('dotenv').config()
  }

  const _wallet = web3.eth.accounts.wallet.add(process.env.ETH_PRIVATE_KEY!)
  // successfully logs the loaded wallet on `yarn generate`
  console.log(_wallet)
  // Issue: `yarn generate` hangs on this line
  return _wallet
}

// initialize the wallet
const wallet = envWallet()

// get the address of the wallet
export const getAddress = () => {
  return wallet.address
}
```

In my root Query, I'm simply calling `getAddress()`. This is all stuff I'm pretty familiar with (even Prisma).
```ts
// ./types/Query.ts
// If I omit this file from the schema, it will compile successfully
import { queryType } from '@nexus/schema'
import { getAddress } from '../modules/Wallet'

console.log('Test 1') // logged after the console.log(_wallet)

const Query = queryType({
  definition(t) {
    // simply a GraphQL query that returns the address of the loaded wallet
    t.string('getAddress', {
      nullable: true,
      resolve: (root, args, ctx) => {
        console.log('Test 2') // never hits
        const address = getAddress() // if I remove this function call, the issue is fixed
        console.log('Test 3') // never hits
        return address
      },
    })
  },
})

export default Query
```

The issue is when I run `yarn generate`, it simply never completes the task. I can see that the wallet is successfully loaded into an object, however, something is causing the Prisma CLI to hang.

# Source
https://github.com/ZaneH/tradr

Follow the steps in the README to run the project. I've made it as easy as I can. Within the `./start.sh` script, I run `yarn generate` and it succeeds because I have commented out the offending lines. You can uncomment them here: https://github.com/ZaneH/tradr/blob/4e66b62980340f3473501ac4e0cb57365ddce425/packages/backend/src/types/Query.ts#L9

Thanks

**Comment by ryands17** [comment]
The only way this could be debugged is if you could share a repro with Web3 as this only happens with this specific lib.

**Reply by ZaneH** [reply]
You are totally right, I've attached the source to the OP: https://github.com/ZaneH/tradr

**Reply by ryands17** [reply]
I cloned your repo, installed deps via `yarn`, went into the _packages/backend_ folder and ran `yarn generate` and it's working fine. I have attached a screenshot below of the same.

![image](https://user-images.githubusercontent.com/19697099/97173828-30327800-17b7-11eb-95b0-7a56c2a935d7.png)

**Reply by ZaneH** [reply]
Sorry, I should have mentioned. Please check the note at the bottom of the OP. I commented out the offending lines so you can get it to compile

**Reply by ryands17** [reply]
![image](https://user-images.githubusercontent.com/19697099/97174938-c61ad280-17b8-11eb-8879-7297bb52c23a.png)

Uncommented and ran `yarn generate` again.

![image](https://user-images.githubusercontent.com/19697099/97174976-d29f2b00-17b8-11eb-8c49-a24133ecaade.png)

Also I don't think this has to do anything with the resolver as the `generate` command in Prisma isn't related to GraphQL. Could you clone this on another system and check?

**Reply by ZaneH** [reply]
I was thinking the same as you (it shouldn't even interact with the resolver) but it seems to have some effect.

I just tried it on a new machine (macOS) and I still can't `yarn generate`:
```sh
~/Downloads/tradr $ yarn generate
yarn run v1.22.5
$ npm -s run generate:prisma && npm -s run generate:nexus
Environment variables loaded from prisma/.env
Prisma Schema loaded from prisma/schema.prisma

✔ Generated Prisma Client (version: 2.9.0) to ./../../node_modules/@prisma/client in 209ms

You can now start using Prisma Client in your code:

import { PrismaClient } from '@prisma/client'
// or const { PrismaClient } = require('@prisma/client')

const prisma = new PrismaClient()

Explore the full API: http://pris.ly/d/client

/Users/zane/Downloads/tradr/node_modules/web3-eth-accounts/lib/index.js:408
    if (!this[account.address]) {
                      ^
TypeError: Cannot read property 'address' of undefined
    at Wallet.add (/Users/zane/Downloads/tradr/node_modules/web3-eth-accounts/lib/index.js:408:23)
    at envWallet (/Users/zane/Downloads/tradr/packages/backend/src/modules/Wallet.ts:14:44)
    at Object.<anonymous> (/Users/zane/Downloads/tradr/packages/backend/src/modules/Wallet.ts:19:16)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Module.m._compile (/Users/zane/Downloads/tradr/node_modules/ts-node/src/index.ts:1043:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/zane/Downloads/tradr/node_modules/ts-node/src/index.ts:1046:12)
    at Module.load (internal/modules/cjs/loader.js:985:32)
    at Function.Module._load (internal/modules/cjs/loader.js:878:14)
    at Module.require (internal/modules/cjs/loader.js:1025:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/Users/zane/Downloads/tradr/packages/backend/src/types/Query.ts:2:1)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Module.m._compile (/Users/zane/Downloads/tradr/node_modules/ts-node/src/index.ts:1043:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/zane/Downloads/tradr/node_modules/ts-node/src/index.ts:1046:12)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

However, I know this is because I'm missing some required environment variables. So I added the environment variables, and I get the exact same issue: I can see the wallet has loaded, but it hangs. I sent you an email containing the environment variables. I don't know why yours didn't crash...........

**Reply by ryands17** [reply]
@ZaneH I got the email with the vars and I could reproduce the issue by running the command from the workspace root:

```
yarn workspace @tradr/backend generate
```

Then it hangs. But if I go to _packages/backend_ and run `yarn generate`, it works fine. So would need to investigate more.

**Reply by ZaneH** [reply]
Wow, mine is the opposite. `yarn workspace ...` gives me a successful compile. Great catch!
edit: It's working now (with this hack), just tested

**Reply by ryands17** [reply]
Strange :)

**Reply by ryands17** [reply, marked as answer]
It's due to `ts-node` transpiling.

If you remove the `generate:nexus` part from this script, it should work:

```
 "generate": "npm -s run generate:prisma && npm -s run generate:nexus",
```
Just run it with Prisma.

**Reply by ZaneH** [reply]
Yup, that makes sense. Something I was not aware of either
Your fix worked, thank you!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
